url4aac7e6f55_simon0191_tetris3d_tgz-pJ8-Tetris3dJ8/src/vsdk/toolkit/gui/CameraControllerAquynza.java

```diff
@@ -0,0 +1,438 @@
+        else if ( val < 0.01 ) val += 0.001;
+        else if ( val < 0.1 - EPSILON ) val += 0.01;
+        else if ( val < 1 - EPSILON ) val += 0.1;
+        else if ( val < 10 - EPSILON ) val += 1;
+        else if ( val < 100 - EPSILON ) val += 10;
+        else if ( val < 1000 - EPSILON ) val += 100;
+        else if ( val < 10000 - EPSILON ) val += 1000;
+        else if ( val < 100000 - EPSILON ) val += 10000;
+        else if ( val < 1000000 - EPSILON ) val += 100000;
+        else if ( val < 10000000 - EPSILON ) val *= 2;
+        else val = 10000000;
+  
+        return val;
+    }
+  
+    private double diminishLogarithmic(double val, double EPSILON)
+    {
+        if ( val > 10000000 + EPSILON ) val /= 2;
+        else if ( val > 1000000 + EPSILON ) val -= 1000000;
+        else if ( val > 100000 + EPSILON ) val -= 100000;
+        else if ( val > 10000 + EPSILON ) val -= 10000;
+        else if ( val > 1000 + EPSILON ) val -= 1000;
+        else if ( val > 100 + EPSILON ) val -= 100;
+        else if ( val > 10 + EPSILON ) val -= 10;
+        else if ( val > 1 + EPSILON ) val -= 1;
+        else if ( val > 0.1 + EPSILON ) val -= 0.1;
+        else if ( val > 0.01 + EPSILON ) val -= 0.01;
+        else if ( val > 0.001 + EPSILON ) val -= 0.001;
+        else if ( val > 0.0001 + EPSILON ) val -= 0.0001;
+        else val = 0.0001;
+        return val;
+    }
+  
+    public boolean processKeyPressedEvent(KeyEvent keyEvent) {
+        // Local copy of the Camera's internal parameters
+        Vector3D eyePosition;
+        Vector3D focusedPosition;
+        Matrix4x4 R; // Camera rotation matrix
+        int projectionMode;
+        double fov;
+        double orthogonalZoom;
+        double nearPlaneDistance;
+        double farPlaneDistance;
+
+        // Internal variables to control the interaction
+        char unicode_id;
+        int keycode;
+        double yaw;
+        double pitch;
+        double roll;
+        double angleInc;
+        boolean updated = false;
+        double EPSILON = 0.0001;
+
+        // 1. Obtain a copy of the camera's internal parameters
+        eyePosition = camera.getPosition();
+        focusedPosition = camera.getFocusedPosition();
+        R = camera.getRotation();
+        projectionMode = camera.getProjectionMode();
+        fov = camera.getFov();
+        orthogonalZoom = camera.getOrthogonalZoom();
+        nearPlaneDistance = camera.getNearPlaneDistance();
+        farPlaneDistance = camera.getFarPlaneDistance();
+
+        // 2. Calculate variables used for interaction manipulation
+        yaw = R.obtainEulerYawAngle();
+        pitch = R.obtainEulerPitchAngle();
+        roll = R.obtainEulerRollAngle();
+  
+        if ( fov > 90 ) angleInc = Math.toRadians(10);
+        else if ( fov > 45 ) angleInc = Math.toRadians(5);
+        else if ( fov > 15 ) angleInc = Math.toRadians(2.5);
+        else if ( fov > 5 ) angleInc = Math.toRadians(1);
+        else angleInc = Math.toRadians(0.1);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 113:117, 386:390.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.